### PR TITLE
[T5] Add 3D attention mask to T5 model (2) (#9643)

### LINKED ISSUE #9643: [Feature Request] Add 3D attention mask for T5Model
## Environment info

- `transformers` version: 4.2.1
- Platform: Linux
- Python version: 3.6.9
- PyTorch version (GPU?): 1.7.0+cuda11.0

### Who can help
T5: @patrickvonplaten

## Information
The `get_extended_attention_mask()` does not sufficiently address the case with 3D attention mask. The problem emerges for T5Model as `input_ids` and `decoder_input_ids` are of different length and the `attention_mask` is of shape [Batch_size, Seq_length, Seq_length]. The decoder uses `attention_mask` directly for `encoder_attention_mask` in cross-attention, which is of incorrect shape and the error message does give any information about why it happens. 

## To reproduce
As described above. I can add code later if needed. 

## Expected behavior
I propose to add a sanity check for attention masks or improve the `get_extended_attention_mask()` method.

### PR DESCRIPTION
**Pull request by lexhuismans**
# What does this PR do?
It allows for 3D attention mask in T5 model (modeling_t5.py) with an accompanying test. 

Fixes #9643

This is a clean version for an earlier PR #10903.

This is a solution for allowing the 3D attention mask in the T5 model by making it broadcastable. It is based on what is used in BERT.

## Who can review?
@patrickvonplaten 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Great job @lexhuismans !

Error is unrelated -> merging

**Review by patrickvonplaten — APPROVED** [review]
